codepropertygraph/src/main/scala/io/shiftleft/passes/CpgPass.scala

```diff
@@ -222,16 +195,74 @@ abstract class ForkJoinParallelCpgPass[T <: AnyRef](cpg: Cpg, outName: String =
         // the nested finally is somewhat ugly -- but we promised to clean up with finish(), we want to include finish()
         // in the reported timings, and we must have our final log message if finish() throws
         val nanosStop = System.nanoTime()
-        val fracRun = if (nanosBuilt == -1) 0.0 else (nanosStop - nanosBuilt) * 100.0 / (nanosStop - nanosStart + 1)
+        val fracRun   = if (nanosBuilt == -1) 0.0 else (nanosStop - nanosBuilt) * 100.0 / (nanosStop - nanosStart + 1)
         val serializationString = if (serializedCpg != null && !serializedCpg.isEmpty) {
           if (inverse) " Inverse serialized and stored." else " Diff serialized and stored."
         } else ""
         baseLogger.info(
-          f"Pass $name completed in ${(nanosStop - nanosStart) * 1e-6}%.0f ms (${fracRun}%.0f%% on mutations). ${nDiff}%d + ${nDiffT - nDiff}%d changes commited from ${nParts}%d parts.${serializationString}%s"
+          f"Pass $name completed in ${(nanosStop - nanosStart) * 1e-6}%.0f ms (${fracRun}%.0f%% on mutations). ${nDiff}%d + ${nDiffT - nDiff}%d changes committed from ${nParts}%d parts.${serializationString}%s"
         )
       }
     }
   }
+
+}
+
+/** NewStyleCpgPassBase is the shared base between ForkJoinParallelCpgPass and ConcurrentWriterCpgPass, containing
+  * shared boilerplate. We don't want ConcurrentWriterCpgPass as a subclass of ForkJoinParallelCpgPass because that
+  * would make it hard to whether an instance is non-racy.
+  *
+  * Please don't subclass this directly. The only reason it's not sealed is that this would mess with our file
+  * hierarchy.
+  */
+abstract class NewStyleCpgPassBase[T <: AnyRef] extends CpgPassBase {
+  type DiffGraphBuilder = overflowdb.BatchedUpdate.DiffGraphBuilder
+  // generate Array of parts that can be processed in parallel
+  def generateParts(): Array[_ <: AnyRef]
+  // setup large data structures, acquire external resources
+  def init(): Unit = {}
```

Comment:
If these need to be define by the child class, then we might just want to leave them empty, as in `def init(): Unit`. That way, the IDE will automatically define them for child classes.

Reply:
`init()` and `finish()` were taken from the now-legacy CpgPass interface -- so you tell me how you intended them to be used ;)

The idea is that a pass may want to run some initialization and cleanup code immediately before and after running. However, most passes don't want to do that, so we do the usual OOP dance of providing a hook in the clazz's vtable that we fill with a NOP default.

Example for `finish()` is closing of file descriptors and clearing pointers to large arrays; example for `init()` is allocation of large arrays and setup code that needs to run serially before the parallel section. Yeah, enterprising individuals could put that into `generateParts()` but better provide a standard way with appropriate guarantees.

The large-array (or generally large-structure) stuff is important because of the way we organize our LayerCreator things: First construct all passes, then run them, and only then drop all references to the passes, allowing them to be garbage collected. Hence, internal shared data structures for parallel passes have an overly large lifetime unless cleared by hand, and hence we want to provide a hook for clearing that stuff.